Add json path to parse error messages

maximkulkin opened the pull request:
Every time you parse some JSON and get error message, it is very painful to track where the error is because existing error messages do not point to the place in JSON where parse error occured. JSON *syntax* errors are reported from Attoparsec parser and it gives position. But JSON *conversion* errors have no support for tracking position/context so quality of error messages is poor.

This change adds json path to error message:
1. Parser tracks current JSON path (in reverse order)
2. An `<?>` operation is added to annotate Parser with JSON context (JSON path). This operation is named after Attoparsec's `<?>` which serves similar purpose.
3. Built-in FromJSON instances are updated to provide context information
4. `.:` and `.:?` operations automatically add context information.

Context is reported in form of [JSON-Path](https://code.google.com/p/json-path/) expression e.g. `$.store.book[0].title`:
```
> eitherDecode storeJsonString :: Either String Store
Left "Error in \"$.store.book[0]\": key \"title\" not present"
```

Caveats:
1. There is no tests for error message yet. I believe those should be done in normal xUnit assertion style instead of QuickCheck property style.
2. Only first error is reported. Further work can include supporting collection of error messages and reporting all of them at once. This can be useful e.g. for validating HTTP JSON requests and reporting all errors at once for displaying them to user on a web form.

zudov commented:
+1
Lack of descriptive error messages is often a pain.
Would like to see it merged.

jsantos17 commented:
@bos I'd love to see this merged. Is there missing here?

maximkulkin commented:
Please note: major2 version will need to be incremented because this patch introduces backward-incompatible changes

dredozubov commented:
I would personally love to see a patch improving these messages merged. Can we revive this issue?

maximkulkin commented:
Rebased to latest master

bos commented on Data/Aeson/Types/Internal.hs:
```diff
@@ -23,10 +23,14 @@ module Data.Aeson.Types.Internal
     -- * Type conversion
     , Parser
     , Result(..)
+    , JsonPathStep(..)
```
I would like to see these named consistently with the rest of the package: `JSON`, not `Json`.

In fact, it would be even better to omit the `Json` or `JSON` prefix entirely as it's redundant.

bos commented on Data/Aeson.hs:
```diff
@@ -149,22 +149,22 @@ decodeStrict' = decodeStrictWith jsonEOF' fromJSON
 -- | Like 'decode' but returns an error message when decoding fails.
 eitherDecode :: (FromJSON a) => L.ByteString -> Either String a
-eitherDecode = eitherDecodeWith jsonEOF fromJSON
+eitherDecode = either (Left . uncurry formatError) Right . eitherDecodeWith jsonEOF fromJSON
```
Two notes:

* These lines are longer than 80 characters.

* The expression `either (Left . uncurry formatError) Right . eitherDecodeWith x y` shows up repeatedly in this file as a result of your diff. That's a strong hint that you should extract it to a function and use that.

bos commented on Data/Aeson/Parser/Internal.hs:
```diff
@@ -30,7 +30,7 @@ module Data.Aeson.Parser.Internal
 import Control.Applicative ((*>), (<$>), (<*), liftA2, pure)
 import Control.Monad.IO.Class (liftIO)
-import Data.Aeson.Types (Result(..), Value(..))
+import Data.Aeson.Types (Result(..), Value(..), JsonPath)
```
I sort names within an import directive alphabetically.

bos commented on Data/Aeson/Parser/Internal.hs:
```diff
@@ -289,27 +289,27 @@ decodeWith p to s =
+    case either (Error []) to (A.parseOnly p s) of
       Success a -> Just a
-      Error _ -> Nothing
+      _         -> Nothing
```
Thank you for fixing the alignment here :-)

bos commented on Data/Aeson/Parser/Internal.hs:
```diff
@@ -289,27 +289,27 @@ decodeWith p to s =
 eitherDecodeWith p to s =
     case L.parse p s of
       L.Done _ v -> case to v of
                       Success a -> Right a
```
Please align the `->` on this line to match the new line you added below.

bos commented on Data/Aeson/Parser/Internal.hs:
```diff
@@ -289,27 +289,27 @@ decodeWith p to s =
 eitherDecodeStrictWith p to s =
-    case either Error to (A.parseOnly p s) of
+    case either (Error []) to (A.parseOnly p s) of
       Success a -> Right a
```
Please match up the `->` :-)

bos commented on Data/Aeson/Types/Instances.hs:
```diff
@@ -101,6 +101,9 @@ import Data.Time.Format (defaultTimeLocale, dateTimeFmt)
 import System.Locale (defaultTimeLocale, dateTimeFmt)
 #endif
 
+parseIndexedJSON :: FromJSON a => Int -> Value -> Parser a
```
Does this need the `JSON` suffix? I think not?

bos commented on Data/Aeson/Types/Instances.hs:
```diff
@@ -553,6 +556,9 @@ instance FromJSON UTCTime where
           _      -> fail "could not parse ISO-8601 date"
     {-# INLINE parseJSON #-}
 
+parseJSONElemAtIndex :: FromJSON a => Int -> Vector Value -> Parser a
```
You could remove the `JSON` infix from this name too.

bos commented on Data/Aeson/Types/Internal.hs:
```diff
@@ -62,25 +66,32 @@ import Data.Vector (Vector)
 import qualified Data.HashMap.Strict as H
 import qualified Data.Vector as V
 
+data JsonPathStep = Key Text | Index Int deriving (Eq, Show)
```
Since you're exposing these new types as part of the public API, they need documentation.

Also, please spread out the declaration of `JsonPathStep` across multiple lines (see `Result` for an indication of how I like to format this).

Finally, `JsonPathStep` should probably be named `Segment`.

bos commented on Data/Aeson/Types/Internal.hs:
```diff
@@ -62,25 +66,32 @@ import Data.Vector (Vector)
               | Success a
                 deriving (Eq, Show, Typeable)
 
+instance NFData JsonPathStep where
```
Please match the indentation to what you see nearby.

bos commented on Data/Aeson/Types/Internal.hs:
```diff
@@ -62,25 +66,32 @@ import Data.Vector (Vector)
     {-# INLINE return #-}
     Success a >>= k = k a
-    Error err >>= _ = Error err
+    Error path err >>= _ = Error path err
```
Please align `>>=`.

bos commented on Data/Aeson/Types/Internal.hs:
```diff
@@ -234,19 +246,28 @@ emptyObject = Object H.empty
+parseEither m v = runParser (m v) [] (\path msg -> Left $ formatError path msg) Right
 {-# INLINE parseEither #-}
 
+formatError :: JsonPath -> String -> String
```
Since you're exposing this via the public API, it needs documentation.

bos commented on Data/Aeson/Types/Internal.hs:
```diff
@@ -256,6 +277,10 @@ object :: [Pair] -> Value
 {-# INLINE object #-}
 
+
+(<?>) :: Parser a -> JsonPathStep -> Parser a
```
Please document this.

bos commented on tests/Properties.hs:
```diff
@@ -58,7 +58,7 @@ encodeInteger i = encode i == L.pack (show i)
 toParseJSON parsejson tojson x =
     case parse parsejson . tojson $ x of
-      Error _ -> False
+      Error _ _ -> False
```
Alignment.

bos commented on tests/Properties.hs:
```diff
@@ -72,7 +72,7 @@ roundTripEq x y = roundTrip (==) x y
 toFromJSON :: (Arbitrary a, Eq a, FromJSON a, ToJSON a) => a -> Bool
 toFromJSON x = case fromJSON . toJSON $ x of
-                Error _ -> False
+                Error _ _ -> False
```
Alignment.

bos commented on tests/Properties.hs:
```diff
@@ -102,7 +102,7 @@ zonedTimeToJSON t = and $
     toFromJSON' :: String -> ZonedTime -> ZonedTime -> Bool
     toFromJSON' format t_ t_' = case fromJSON . toJSON' format $ t_ of
-                                  Error msg -> error' msg
+                                  Error _ msg -> error' msg
```
Alignment.

bos commented:
Overall, I think I like this.

You can see that I've looked through the code and have asked for a number of insignificant formatting changes.

Before you go tweaking the formatting, please measure the impact on performance.

To do that, run some of the benchmarks before and after your change, and post a comment here (and update the commit description) to describe what difference you measure:

```
cd benchmarks
cabal sandbox init
cabal install -j --only-dependencies
cabal configure
cabal build -j
dist/build/aeson-benchmark-compare-with-json/aeson-benchmark-compare-with-json -m glob '**/*aeson*'
```

maximkulkin commented:
Benchmark without patch:
```
benchmarking decode/en/aeson
time                 1.793 ms   (1.762 ms .. 1.824 ms)
                     0.997 R²   (0.994 R² .. 0.998 R²)
mean                 1.797 ms   (1.771 ms .. 1.832 ms)
std dev              98.14 μs   (74.03 μs .. 138.2 μs)
variance introduced by outliers: 39% (moderately inflated)

benchmarking decode/jp/aeson
time                 2.475 ms   (2.428 ms .. 2.511 ms)
                     0.997 R²   (0.995 R² .. 0.998 R²)
mean                 2.550 ms   (2.516 ms .. 2.602 ms)
std dev              133.7 μs   (93.08 μs .. 202.7 μs)
variance introduced by outliers: 36% (moderately inflated)

benchmarking encode/en/aeson-to-bytestring
time                 413.0 μs   (403.6 μs .. 424.0 μs)
                     0.996 R²   (0.994 R² .. 0.999 R²)
mean                 408.3 μs   (403.3 μs .. 414.8 μs)
std dev              18.40 μs   (14.98 μs .. 24.01 μs)
variance introduced by outliers: 39% (moderately inflated)

benchmarking encode/en/aeson-via-text-to-bytestring
time                 1.073 ms   (1.054 ms .. 1.097 ms)
                     0.995 R²   (0.990 R² .. 0.998 R²)
mean                 1.067 ms   (1.052 ms .. 1.091 ms)
std dev              59.81 μs   (40.56 μs .. 95.40 μs)
variance introduced by outliers: 45% (moderately inflated)

benchmarking encode/en/aeson-to-text
time                 929.9 μs   (911.6 μs .. 948.0 μs)
                     0.997 R²   (0.995 R² .. 0.998 R²)
mean                 926.7 μs   (913.2 μs .. 942.7 μs)
std dev              51.20 μs   (42.51 μs .. 62.28 μs)
variance introduced by outliers: 46% (moderately inflated)

benchmarking encode/jp/aeson-to-bytestring
time                 410.2 μs   (405.5 μs .. 414.7 μs)
                     0.998 R²   (0.997 R² .. 0.999 R²)
mean                 409.5 μs   (404.6 μs .. 414.4 μs)
std dev              16.14 μs   (13.74 μs .. 18.77 μs)
variance introduced by outliers: 34% (moderately inflated)

benchmarking encode/jp/aeson-via-text-to-bytestring
time                 1.095 ms   (1.069 ms .. 1.118 ms)
                     0.996 R²   (0.994 R² .. 0.997 R²)
mean                 1.099 ms   (1.080 ms .. 1.126 ms)
std dev              74.67 μs   (57.21 μs .. 112.3 μs)
variance introduced by outliers: 55% (severely inflated)

benchmarking encode/jp/aeson-to-text
time                 996.9 μs   (947.6 μs .. 1.039 ms)
                     0.989 R²   (0.984 R² .. 0.994 R²)
mean                 936.3 μs   (919.3 μs .. 958.8 μs)
std dev              64.58 μs   (51.88 μs .. 82.76 μs)
variance introduced by outliers: 56% (severely inflated)
```

Benchmark with patch:
```
benchmarking decode/en/aeson
time                 1.792 ms   (1.751 ms .. 1.849 ms)
                     0.997 R²   (0.995 R² .. 0.999 R²)
mean                 1.777 ms   (1.759 ms .. 1.799 ms)
std dev              65.93 μs   (54.43 μs .. 81.57 μs)
variance introduced by outliers: 23% (moderately inflated)

benchmarking decode/jp/aeson
time                 2.533 ms   (2.479 ms .. 2.603 ms)
                     0.995 R²   (0.993 R² .. 0.997 R²)
mean                 2.565 ms   (2.528 ms .. 2.613 ms)
std dev              133.9 μs   (106.6 μs .. 169.1 μs)
variance introduced by outliers: 34% (moderately inflated)

benchmarking encode/en/aeson-to-bytestring
time                 413.5 μs   (407.1 μs .. 418.3 μs)
                     0.998 R²   (0.997 R² .. 0.999 R²)
mean                 412.5 μs   (406.9 μs .. 421.3 μs)
std dev              22.61 μs   (15.53 μs .. 33.70 μs)
variance introduced by outliers: 50% (moderately inflated)

benchmarking encode/en/aeson-via-text-to-bytestring
time                 1.115 ms   (1.092 ms .. 1.138 ms)
                     0.997 R²   (0.996 R² .. 0.998 R²)
mean                 1.098 ms   (1.085 ms .. 1.115 ms)
std dev              49.12 μs   (41.44 μs .. 57.85 μs)
variance introduced by outliers: 33% (moderately inflated)

benchmarking encode/en/aeson-to-text
time                 909.1 μs   (895.2 μs .. 928.1 μs)
                     0.996 R²   (0.994 R² .. 0.998 R²)
mean                 938.7 μs   (922.2 μs .. 956.4 μs)
std dev              59.11 μs   (50.91 μs .. 70.45 μs)
variance introduced by outliers: 51% (severely inflated)

benchmarking encode/jp/aeson-to-bytestring
time                 424.9 μs   (418.2 μs .. 432.7 μs)
                     0.997 R²   (0.995 R² .. 0.998 R²)
mean                 435.4 μs   (427.8 μs .. 448.3 μs)
std dev              31.43 μs   (20.61 μs .. 52.74 μs)
variance introduced by outliers: 63% (severely inflated)

benchmarking encode/jp/aeson-via-text-to-bytestring
time                 1.103 ms   (1.082 ms .. 1.126 ms)
                     0.997 R²   (0.996 R² .. 0.999 R²)
mean                 1.108 ms   (1.092 ms .. 1.130 ms)
std dev              62.74 μs   (48.19 μs .. 89.31 μs)
variance introduced by outliers: 45% (moderately inflated)

benchmarking encode/jp/aeson-to-text
time                 929.5 μs   (915.1 μs .. 943.1 μs)
                     0.998 R²   (0.996 R² .. 0.999 R²)
mean                 924.3 μs   (911.5 μs .. 944.5 μs)
std dev              52.88 μs   (34.71 μs .. 74.71 μs)
variance introduced by outliers: 47% (moderately inflated)
```

maximkulkin commented:
I have addressed all issues except for:
* `JsonPath` and `JsonPathStep` were renamed to `JSONPath` and `JSONPathElement`. I feel like stripping JSON prefix would make JSONPath look like a regular file system path that would confuse users.
* `parseJSONIndexed` and `parseJSONElemAtIndex` I left unchanged. The reason for JSON infix there is to mimic regular `parseJSON` (which in fact they are with a little additions)

Also I have added "Since 0.9" documentation to `<?>` function. It needed such annotation but real version should be tweaked.

Benchmarks show a slight (IMO insignificant) slow down on benchmarks. Assuming the benefits this patch provides I think it is worth it.

maximkulkin commented:
Here is a demo:
```
-- Person.hs
module Person where

import Control.Applicative ((<$>), (<*>))
import Data.Aeson

data Address = Address
  { addressStreet :: String
  , addressCountry :: String
  } deriving (Eq, Show)

data Person = Person
  { personName :: String
  , personAge :: Int
  , personAddress :: Address
  } deriving (Eq, Show)

instance FromJSON Address where
  parseJSON = withObject "Address" $ \o ->
    Address
      <$> o .: "street"
      <*> o .: "country"

instance FromJSON Person where
  parseJSON = withObject "Person" $ \o ->
    Person
      <$> o .: "name"
      <*> o .: "age"
      <*> o .: "address"
```
Save given text into Person.hs file, then
```
cabal repl
:set -XOverloadedStrings
:load Person.hs
*Person> import Data.Aeson
*Person Data.Aeson> eitherDecode "{\"name\": \"John\"}" :: Either String Person
Left "Error in $: key \"age\" not present"
*Person Data.Aeson> eitherDecode "{\"name\": \"John\", \"age\": 32}" :: Either String Person
Left "Error in $: key \"address\" not present"
*Person Data.Aeson> eitherDecode "{\"name\": \"John\", \"age\": 32, \"address\": {}}" :: Either String Person
Left "Error in $.address: key \"street\" not present"
*Person Data.Aeson> eitherDecode "{\"name\": \"John\", \"age\": 32, \"address\": {\"street\": 123}}" :: Either String Person
Left "Error in $.address.street: when expecting a String, encountered Number instead"
*Person Data.Aeson> eitherDecode "{\"name\": \"John\", \"age\": 32, \"address\": {\"street\": \"Foo\"}}" :: Either String Person
Left "Error in $.address: key \"country\" not present"
*Person Data.Aeson> eitherDecode "{\"name\": \"John\", \"age\": 32, \"address\": {\"street\": \"Foo\", \"country\": \"Bar\"}}" :: Either String Person
Right (Person {personName = "John", personAge = 32, personAddress = Address {addressStreet = "Foo", addressCountry = "Bar"}})
*Person Data.Aeson> eitherDecode "[{\"name\": \"John\", \"age\": 32, \"address\": {\"street\": \"Foo\", \"country\": 123}}]" :: Either String [Person]
Left "Error in $[0].address.country: when expecting a String, encountered Number instead"
```

bos commented:
Thanks, @maximkulkin! Very nice work.

bitemyapp commented:
@bos @maximkulkin this looks _really_ nice, when could it land on Hackage? (AFAICT, it's not in 0.9.0.1)

I can vendor in the meantime but having it up would be :+1:

bos commented:
@bitemyapp I just rebased to the 0.9 branch. Could you please give it a try and see if it introduces any breaking changes for you?

bitemyapp commented:
@bos I need to start a branch to migrate off the deprecated Generics, but once I do, I'll let you know if it breaks anything.

bos commented:
Cool, thanks. I'll try a few downstream packages myself and see if they work.
